data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwarderReceiveBuffer.java

```diff
@@ -0,0 +1,127 @@
+            }
+            blockingQueue.offer(record);
+        } catch (InterruptedException ex) {
+            LOG.error("Plugin [{}] - Buffer is full, interrupted while waiting to write the record", pluginName, ex);
+            throw new TimeoutException("Buffer is full, timed out waiting for a slot");
+        }
+    }
+
+    @Override
+    public void writeAll(Collection<T> records, int timeoutInMillis) throws Exception {
+        final int size = records.size();
+        if (size > bufferSize) {
+            throw new SizeOverflowException(format("Buffer capacity too small for the size of records: %d", size));
+        }
+        try {
+            final boolean permitAcquired = capacitySemaphore.tryAcquire(size, timeoutInMillis, TimeUnit.MILLISECONDS);
+            if (!permitAcquired) {
+                throw new TimeoutException(
+                        format("Plugin [%s] - Buffer does not have enough capacity left for the size of records: %d, " +
+                                        "timed out waiting for slots.",
+                                pluginName, size));
+            }
+            blockingQueue.addAll(records);
+        } catch (InterruptedException ex) {
+            LOG.error("Plugin [{}] - Buffer does not have enough capacity left for the size of records: {}, " +
+                            "interrupted while waiting to write the records",
+                    pluginName, size, ex);
+            throw new TimeoutException(
+                    format("Plugin [%s] - Buffer does not have enough capacity left for the size of records: %d, " +
+                                    "timed out waiting for slots.",
+                            pluginName, size));
+        }
+    }
+
+    @Override
+    public Map.Entry<Collection<T>, CheckpointState> read(int timeoutInMillis) {
+        final List<T> records = new ArrayList<>();
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        try {
+            while (stopwatch.elapsed(TimeUnit.MILLISECONDS) < timeoutInMillis && records.size() < batchSize) {
+                final T record = blockingQueue.poll(timeoutInMillis, TimeUnit.MILLISECONDS);
+                if (record != null) { //record can be null, avoiding adding nulls
+                    records.add(record);
+                }
+                if (records.size() < batchSize) {
+                    blockingQueue.drainTo(records, batchSize - records.size());
+                }
+            }
+        } catch (InterruptedException ex) {
+            LOG.info("Plugin [{}] - Interrupt received while reading from buffer", pluginName);
+            throw new RuntimeException(ex);
+        }
+        final CheckpointState checkpointState = new CheckpointState(records.size());
+        recordsInFlight += checkpointState.getNumRecordsToBeChecked();
+        return new AbstractMap.SimpleEntry<>(records, checkpointState);
+    }
+
+    @Override
+    public void checkpoint(CheckpointState checkpointState) {
```

Comment:
QUES: Do we need checkpointing in core peerforwarder buffer? If yes, when will this API be invoked? Asking since the buffer capacity semaphore will not be released unless this API is invoked properly.

Reply:
This will be invoked after reading events from buffer by the `PeerForwardingProcessorDecorator` and sent it to other peers or sent to process locally.